### PR TITLE
Fix problem with String.equals not checking the argument is a String TG-1619

### PR DESCRIPTION
**Pull request by romainbrenguier**
The previous implementation was assuming the given Object was a String, which could lead CBMC into mistakenly thinking objects of other types (Integer for instance) could be equal to a string.

### CODE REVIEW

**Review by allredj — APPROVED** [review]
Looks fine, as far as I can tell.

File: src/java_bytecode/java_object_factory.cpp [hunk 1]
```diff
@@ -571,9 +571,15 @@ codet initialize_nondet_string_struct(
 
   // `obj` is `*expr`
   const struct_typet &struct_type = to_struct_type(ns.follow(obj.type()));
-  const irep_idt &class_id = "java::" + id2string(struct_type.get_tag());
-
-  // @clsid = String and @lock = false:
+  // @clsid = java::java.lang.String or similar
+  // Note that CharSequence cannot be used as classid because it's abstract
+  // \todo allow StringBuffer and StringBuilder as classid for Charsequence
+  const irep_idt &class_id =
```
Comment (smowton) [inline src/java_bytecode/java_object_factory.cpp]: Explain in the comment why you're allowing classes other than `String` to be aliased with `String` here?

File: src/java_bytecode/java_string_library_preprocess.cpp [hunk 2]
```diff
@@ -923,8 +922,30 @@ codet java_string_library_preprocesst::make_equals_function_code(
...
+  const symbolt string_equals_sym = get_fresh_aux_symbol(
+    java_boolean_type(), "string_equals", "str_eq", loc, ID_java, symbol_table);
+  const symbol_exprt string_equals = string_equals_sym.symbol_expr();
+  code.add(code_declt(string_equals));
+  code.add(code_assignt(
+    string_equals,
+    make_function_application(
+      ID_cprover_string_equal_func, args, type.return_type(), symbol_table)));
+
+  code.add(code_returnt(if_exprt(
```
Comment (smowton) [inline src/java_bytecode/java_string_library_preprocess.cpp:945]: `and_exprt`?

Comment (romainbrenguier) [reply src/java_bytecode/java_string_library_preprocess.cpp:945]: The problem is that they don't have the same type: `arg_is_string` has `bool_type()` and `string_equals` has `java_boolean_type()`, these have different bitwidth, What I want in the end is something of type `java_boolean_type()`

Comment (smowton) [reply src/java_bytecode/java_string_library_preprocess.cpp:945]: fair enough

**Review by smowton — APPROVED** [review]

**Pull request merged**